### PR TITLE
Dotnet 2.1 and FAKE 5

### PR DESCRIPTION
**Pull request by theimowski**

### CODE REVIEW

**Comment by isaacabraham** [comment]
This looks good. I wonder if it fixes that big with the azure deploy where you need to save the file to get it to work

**Comment by theimowski** [comment]
right, might be worth looking at but then I'm not sure how to check it if it occurred only first time

**Comment by theimowski** [comment]
@matthid can you please help reviewing the changes?

I'm wondering what'd be the best UX for someone creating project from the template and then running the `run` FAKE 5 target.

Currently it's a couple of steps after `dotnet new SAFE` from scratch:

1. .paket\paket.exe install (this is due to https://github.com/SAFE-Stack/SAFE-template/issues/66)
1. dotnet restore build.proj
1. dotnet fake build target run

How can we minimize the number of steps?

**Comment by matthid** [comment]
@theimowski One easy win is to write the mentioned steps into msbuild and run `dotnet build build.proj`, which trims it down to a single command.

If you want you can recommend people to install fake via global tools or package manager to type less (and be a bit faster because no implicit restore is required)

File: Content/build.fsx [hunk 1]
```diff
@@ -1,4 +1,5 @@
-#r @"packages/build/FAKE/tools/FakeLib.dll"
+#r "paket: groupref build //"
+#load "./.fake/build.fsx/intellisense.fsx"
 //#if (deploy == "azure")
 #r "netstandard"
 #I "packages/build/Microsoft.Rest.ClientRuntime.Azure/lib/net452"
```
[the remark this answers is not in the document]
Comment (isaacabraham) [reply Content/build.fsx]: The #I used to be needed for FSI to get the correct dependencies but with Fake 5 maybe it's no longer needed.

**Comment by theimowski** [comment]
Good idea with paket groups - haven't considered that. The issue though would be that original paket.lock would need to have plenty of conditionals. Plus the number of groups would be quite big.

> install fake via global tools or package manager

wondering whether this dependency could potentially discourage new folks?

**Comment by theimowski** [comment]
with the global tools it boils down to just single command `fake run build.fsx --target run` after `dotnet new SAFE` - it runs `paket install` if paket.lock was not found

File: Content/build.fsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (theimowski) [reply Content/build.fsx]: How about paket-files? does `#r paket: groupref build //" support loading those as well?

**Comment by Krzysztof-Cieslak** [comment]
I think global tools are used by basically any other ecosystem I’ve tried

**Comment by matthid** [comment]
@theimowski What I meant was that:

- All you need is `dotnet build`. It will restore `dotnet-fake` and run it (easy bootstrap)
- Once you are familiar and when working locally it makes a lot more sense to just install the global tool (regular workflow)

**Comment by theimowski** [comment]
I think I like the option with global tool - no `build.proj`, no `build.cmd` / `build.sh` just `fake run build.fsx`

**Comment by theimowski** [comment]
@isaacabraham  I'm getting a load error in ArmTemplate target. Can this be due to `github CompositionalIT/fshelpers src/FsHelpers/ArmHelper/ArmHelper.fs` as a file dependency?

```
---------------------------------------------------------------------
Build Time Report
---------------------------------------------------------------------
Target          Duration
------          --------
Clean           00:00:00.1472840
InstallClient   00:00:02.8825805
Build           00:00:21.7015925
Bundle          00:00:05.1646290
ArmTemplate     00:00:00.0114603   (Could not load type 'Microsoft.FSharp.Control.AsyncActivation`1' from assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.)
AppService      00:00:00           (skipped)
Total:          00:00:30.0309497
Status:         Failure
---------------------------------------------------------------------
Script reported an error:
-> BuildFailedException: Target 'ArmTemplate' failed.
-> One or more errors occurred. (Could not load type 'Microsoft.FSharp.Control.AsyncActivation`1' from assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.)
-> TypeLoadException: Could not load type 'Microsoft.FSharp.Control.AsyncActivation`1' from assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

**Comment by isaacabraham** [comment]
Weird, not sure about that. I don't even recognise that type. I'm back home on the weekend so will have a look next week. If you compile the template in fsi does it give the same error?

**Comment by matthid** [comment]
Looks like a potential breaking change in FSharp.Core. Should probably be reported.

**Comment by matthid** [comment]
Or you are using an older version of FSharp.Core.

**Comment by theimowski** [comment]
It works fine in FSI. I tried to Set COREHOST_TRACE=1 and see where it's coming from, but nothing interesting... Locked version is FSharp.Core 4.5 

Should i report it on visualfsharp?

**Comment by matthid** [comment]
@theimowski Can you try the suggested workaround in https://github.com/fsharp/FAKE/issues/2001 and pin the FSharp.Core version in the fake dependencies?

**Comment by matthid** [comment]
Also please leave feedback on the fake issue what you would have expected from a users perspective in this scenario. I have a favorite solution but I'd like to hear what you think

**Comment by theimowski** [comment]
thanks @matthid, pinning FSharp.Core did the job.

@isaacabraham it deploys to azure now fine, but there are warnings to ArmTemplate target:

```
~\sandbox\test\paket-files/build/CompositionalIT/fshelpers/src/FsHelpers/ArmHelper/ArmHelper.fs (17,9)-(17,22): Warning FS1178: The struct, record or union type 'ParameterType' is not structurally comparable because the type 'obj' does not satisfy the 'comparison' constraint. Consider adding the 'NoComparison' attribute to the type 'ParameterType' to clarify that the type is not comparable
~\sandbox\test\paket-files/build/CompositionalIT/fshelpers/src/FsHelpers/ArmHelper/ArmHelper.fs (31,5)-(31,22): Warning FS1178: The struct, record or union type 'ResourceGroupType' is not structurally comparable because the type 'Core.Region' does not satisfy the 'comparison' constraint. Consider adding the 'NoComparison' attribute to the type 'ResourceGroupType' to clarify that the type is not comparable
~\sandbox\test\paket-files/build/CompositionalIT/fshelpers/src/FsHelpers/ArmHelper/ArmHelper.fs (33,5)-(33,15): Warning FS1178: The struct, record or union type 'Deployment' is not structurally comparable because the type 'ResourceGroupType' does not satisfy the 'comparison' constraint. Consider adding the 'NoComparison' attribute to the type 'Deployment' to clarify that the type is not comparable
~\sandbox\test\paket-files/build/CompositionalIT/fshelpers/src/FsHelpers/ArmHelper/ArmHelper.fs (39,5)-(39,25): Warning FS1178: The struct, record or union type 'AuthenticatedContext' is not structurally comparable because the type 'IResourceManager' does not satisfy the 'comparison' constraint. Consider adding the 'NoComparison' attribute to the type 'AuthenticatedContext' to clarify that the type is not comparable
~\sandbox\test\build.fsx (111,69)-(111,94): Warning FS0052: The value has been copied to ensure the original is not mutated by this operation or because the copy is implicit when returning a struct from a member and another member is then accessed
```

**Comment by isaacabraham** [comment]
@theimowski I assume that these are new warnings in latest F# compiler or F# core or something?

**Comment by theimowski** [comment]
haven't spotted them before so not sure if they're new

**Comment by theimowski** [comment]
This is now ready to be merged, tested azure and docker integration + other basic template options

@isaacabraham @matthid please review and I'll release

**Comment by isaacabraham** [comment]
@theimowski can you hold fire for a day or two for the review? If not, go ahead and release - I'm sure it's fine if you've tested it (note - I have a SAFE talk tomorrow where I'll be showing the Azure release process :-))

**Comment by isaacabraham** [comment]
@forki you know me so well.

**Comment by theimowski** [comment]
Sure thing, I'll hold on

I know that feeling, the github TLS 1.3 paket issue occurred just before I did my first SAFE talk at LambdaDays earlier this year.
Thankfully I discovered that before the talk and managed to apply a workaround ;)

**Comment by isaacabraham** [comment]
BTW - just tried the old template with FAKE 5 - still the same issue with the "can't find dependency" stuff for Azure. Maybe with this upgrade it'll magically go away...

**Comment by matthid** [comment]
@isaacabraham it will go away if you lock fsharp.core or when we upgrade fake to 4.5

**Comment by theimowski** [comment]
I think @isaacabraham meant https://github.com/SAFE-Stack/SAFE-template/pull/65#issuecomment-385619990

**Comment by isaacabraham** [comment]
@theimowski just testing out locally, seems to work nicely! The only things I've noticed is more around documentation - I didn't know how to launch without a build.cmd (instead it's now `fake build run`?), so perhaps we could retain a build.cmd that just does that, or include some docs on that?

**Comment by theimowski** [comment]
It's `fake run build.fsx [--target run]` - i added https://github.com/SAFE-Stack/docs/pull/51 for this change

**Pull request merged**